### PR TITLE
Evasion attack on image classification model

### PR DESCRIPTION
**Pull request by harimohanr**
Comprehensive vulnerability assessment for evasion attack on machine learning classifiers trained on MNIST dataset. 

### Key Features:
- **Multi-Attack Support**: Implements 15+ adversarial attack methods including FGSM, PGD, BIM, DeepFool, C&W, and AutoAttack
- **Flexible Attack Modes**: Three predefined attack suites (fast, comprehensive, custom) with configurable parameters
- **Binary & Multi-class Classification**: Supports both binary and 10-class MNIST classification scenarios
- **TensorFlow/Keras Integration**: Compatible with TensorFlow 2.x models using ART (Adversarial Robustness Toolbox)
- **Comprehensive Reporting**: Generates detailed attack results with accuracy metrics, perturbation statistics (L∞, L2), and success rates
- **Visualization**: Creates color-coded bar charts showing model vulnerability across different attacks
- **CLI Interface**: Command-line tool with configurable parameters for batch testing

### Technical Highlights:
- Robust error handling with fallback mechanisms for classifier creation
- JSON export of results with metadata for reproducibility  
- Automatic model compilation and validation
- Memory-efficient sample-based testing
- Cross-platform compatibility

### Use Cases:
- Security assessment of ML models before deployment
- Adversarial training validation
- Model robustness benchmarking
- Red team exercises for AI systems

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

### Test Configuration
- **Environment**: macOS Python 3.11.4, pytest 8.4.1
- **Dependencies**: TensorFlow 2.x, ART, Keras, NumPy
- **Test Suite**: 10 test classes, 22 individual test cases

#### Unit Tests (18 tests)
- **Model Loading**: File validation, compilation, error handling
- **Data Preparation**: MNIST processing, binary classification filtering, normalization
- **ART Classifier Creation**: TensorFlow/Keras integration with fallback mechanisms
- **Attack Execution**: Success/failure scenarios, perturbation calculations
- **Configuration Management**: Attack parameter validation, JSON serialization
- **I/O Operations**: Results saving with numpy type conversion

#### Integration Tests (4 tests)
- **End-to-End Workflow**: Real Keras model with Conv2D architecture
- **ModelWrapper**: Prediction validation with softmax outputs
- **Attack Pipeline**: Complete adversarial example generation

### Manual Verification
```bash
python rt_binary_classifier.py --model-path model.h5 --attack-type fast --num-samples 50 --visualize
```
**Verified**: Model loading, 4 adversarial attacks (FGSM, BIM, PGD), results export, visualization generation

### Results
- ✅ **22/22 tests passed** 
- ✅ All error handling scenarios covered
- ✅ TensorFlow eager execution compatibility resolved
- ✅ ART library integration validated
- ✅ Cross-platform file operations confirmed

### Key Fixes Validated
- Import path resolution for test modules
- TensorFlow/ART classifier fallback mechanisms  
- Numpy type serialization for JSON export
- Graceful degradation for missing dependencies

All functionality verified working correctly in both isolated unit tests and integrated workflows.

# Checklist

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] My changes have been tested using both CPU and GPU devices

### CODE REVIEW

**Comment by harimohanr** [comment]
@panickervinod Please review this PR

**Comment by panickervinod** [comment]
@harimohanr  do refactor the file names and evaluate if the model creation code must be in the resources folder
